src/main.rs

```diff
@@ -31,81 +33,138 @@ struct Kopium {
-    let client = Client::try_default().await?;
-    let api: Api<CustomResourceDefinition> = Api::all(client);
-    let crd = api.get(&kopium.crd).await?;
+    Kopium::from_args().dispatch().await
+}
 
-    let version = find_crd_version(&crd, kopium.api_version.as_deref())?;
-    let data = version
-        .schema
-        .as_ref()
-        .and_then(|schema| schema.open_api_v3_schema.clone());
-    let version = version.name.clone();
+impl Kopium {
+    async fn dispatch(&self) -> Result<()> {
+        let crds = Client::try_default().await.map(Crds::all)?;
 
-    let kind = crd.spec.names.kind;
-    let plural = crd.spec.names.plural;
-    let group = crd.spec.group;
-    let scope = crd.spec.scope;
+        if let Some(name) = self.crd.as_deref() {
+            if self.command.is_none() {
+                self.generate(crds, name).await?;
+            } else {
+                self.help()?;
+            }
+        } else {
+            match self.command {
+                Some(Command::ListCrds) => self.list_crds(crds).await?,
+                Some(Command::Completions { shell }) => self.completions(shell)?,
+                None => self.help()?,
+            }
+        }
 
+        Ok(())
+    }
 
-    if let Some(schema) = data {
-        let mut results = vec![];
-        log::debug!("schema: {}", serde_json::to_string_pretty(&schema)?);
-        analyze(schema, "", &kind, 0, &mut results)?;
+    fn help(&self) -> Result<()> {
+        Self::clap().print_help().map(|_| println!())?;
+        Ok(())
+    }
 
-        if !kopium.hide_prelude {
-            print_prelude(&results);
-        }
+    async fn list_crds(&self, crds: Crds) -> Result<()> {
+        let lp = api::ListParams::default();
+        crds.list(&lp).await?.items.iter().for_each(|crd| {
+            println!("{}", crd.name());
+        });
+        Ok(())
+    }
 
-        for s in results {
-            if s.level == 0 {
-                continue; // ignoring root struct
-            } else {
-                if s.level == 1 && s.name.ends_with("Spec") {
-                    print_derives(&kopium.derive);
-                    println!(
-                        r#"#[kube(group = "{}", version = "{}", kind = "{}", plural = "{}")]"#,
-                        group, version, kind, plural
-                    );
-                    if scope == "Namespaced" {
-                        println!(r#"#[kube(namespaced)]"#);
-                    }
-                    // don't support grabbing original schema atm so disable schemas:
-                    // (we coerce IntToString to String anyway so it wont match anyway)
-                    println!(r#"#[kube(schema = "disabled")]"#);
-                    println!("pub struct {} {{", s.name);
+    fn completions(&self, shell: clap::Shell) -> Result<()> {
+        use std::io::Write;
+        let mut stdout = std::io::stdout();
+        Self::clap().gen_completions_to("kopium", shell, &mut stdout);
+        writeln!(stdout)?;
```

Comment:
Are you using `Write` here to make it pipeable? (i.e. does it work if you do `kopium completions | head` ?)
I'm used to just using `print!("{}", stdout)` and setting `unsafe { libc::signal(libc::SIGPIPE, libc::SIG_DFL); }` but this is fine due to the SIGPIPE bug in rust.

This isn't a big deal, but ought to be consistent with the rest of the writing (which is using `print` / `println`).

Reply:
Well, in fact, it is due to `gen_completion_to` emitting its output to a `Write` object, I guess I can use `String` instead and then `println!` it.

Reply:
Turns out to be less nice than I thought, `String` is not `Write`, so I use `Vec<u8>` instead